docs/wingetopt.h

```diff
@@ -0,0 +1,671 @@
+	if (*options == '-')
+		flags |= FLAG_ALLARGS;
+	else if (posixly_correct || *options == '+')
+		flags &= ~FLAG_PERMUTE;
+	if (*options == '+' || *options == '-')
+		options++;
+
+	optarg = NULL;
+	if (optreset)
+		nonopt_start = nonopt_end = -1;
+start:
+	if (optreset || !*place) {                /* update scanning pointer */
+		optreset = 0;
+		if (optind >= nargc) { /* end of argument vector */
+			place = EMSG;
+			if (nonopt_end != -1) {
+				/* do permutation, if we have to */
+				permute_args(nonopt_start, nonopt_end,
+					optind, nargv);
+				optind -= nonopt_end - nonopt_start;
+			}
+			else if (nonopt_start != -1) {
+				/*
+				* If we skipped non-options, set optind
+				* to the first of them.
+				*/
+				optind = nonopt_start;
+			}
+			nonopt_start = nonopt_end = -1;
+			return (-1);
+		}
+		if (*(place = nargv[optind]) != '-' ||
+			(place[1] == '\0' && strchr(options, '-') == NULL)) {
+			place = EMSG;                /* found non-option */
+			if (flags & FLAG_ALLARGS) {
+				/*
+				* GNU extension:
+				* return non-option as argument to option 1
+				*/
+				optarg = nargv[optind++];
+				return (INORDER);
+			}
+			if (!(flags & FLAG_PERMUTE)) {
+				/*
+				* If no permutation wanted, stop parsing
+				* at first non-option.
+				*/
+				return (-1);
+			}
+			/* do permutation */
+			if (nonopt_start == -1)
+				nonopt_start = optind;
+			else if (nonopt_end != -1) {
+				permute_args(nonopt_start, nonopt_end,
+					optind, nargv);
+				nonopt_start = optind -
+					(nonopt_end - nonopt_start);
+				nonopt_end = -1;
+			}
+			optind++;
+			/* process next argument */
+			goto start;
+		}
+		if (nonopt_start != -1 && nonopt_end == -1)
+			nonopt_end = optind;
+
+		/*
+		* If we have "-" do nothing, if "--" we are done.
+		*/
+		if (place[1] != '\0' && *++place == '-' && place[1] == '\0') {
+			optind++;
+			place = EMSG;
+			/*
+			* We found an option (--), so if we skipped
+			* non-options, we have to permute.
+			*/
+			if (nonopt_end != -1) {
+				permute_args(nonopt_start, nonopt_end,
+					optind, nargv);
+				optind -= nonopt_end - nonopt_start;
+			}
+			nonopt_start = nonopt_end = -1;
+			return (-1);
+		}
+	}
+
+	/*
+	* Check long options if:
+	* 1) we were passed some
+	* 2) the arg is not just "-"
+	* 3) either the arg starts with -- we are getopt_long_only()
+	*/
+	if (long_options != NULL && place != nargv[optind] &&
+		(*place == '-' || (flags & FLAG_LONGONLY))) {
+		short_too = 0;
+		if (*place == '-')
+			place++;                /* --foo long option */
+		else if (*place != ':' && strchr(options, *place) != NULL)
+			short_too = 1;                /* could be short option too */
+
+		optchar = parse_long_options(nargv, options, long_options,
+			idx, short_too);
+		if (optchar != -1) {
+			place = EMSG;
+			return (optchar);
+		}
+	}
+
+	if ((optchar = (int)*place++) == (int)':' ||
+		(optchar == (int)'-' && *place != '\0') ||
+		(oli = strchr(options, optchar)) == NULL) {
```

Comment:
VS is giving compilation issues on this line saying that a const char * cant be cast to a char *:
![image](https://user-images.githubusercontent.com/70495470/205176621-1a9c66e8-8708-46d0-8519-bf3bf4ae4dd6.png)

It seems that VS is preferring the `const` overload of [`strchr`](https://cplusplus.com/reference/cstring/strchr/) rather than the non-const version.

The new `wingetopt.h` is working for me on WSL and the 281 AG.

Reply:
All I did was copy-paste your `getopt.c` into your `getop.h` at the end and then rename it.  Did you have the same problem with two file version?

Reply:
I assumed that that is all you did. I will try to repro on a CAEN computer as well.

I didn't have this is issue with the two file version.

Reply:
Hmm.  Here's a hypothesis.  The files you published are different than whatever is living in your system's copy of `getopt.h`.  Another thing to try.  Rename `getopt.h` -> `xgetopt.h`, `getopt.c` -> `xgetopt.c` and then `#include "xgetopt.h`.  That will for Visual Studio to use the code provided.  See if that reproduces the problem.

Reply:
Another hypothesis: the old `getopt.h` installed on your system has the *same include guards*.  Maybe your project does `#include <getopt.h>` elsewhere.  We should customize the include guards to be `__WIN_GETOPT` or something like that.

Reply:
I can try renaming it, but I don't think that would fix any of the issues. The VS project wouldn't know of the existence of the old `getopt.h`, as the new file we are including is `wingetopt.h`

Reply:
I am also having the same issue on CAEN (which I assume is just a fresh install of VS).

We also need to comment out `#include <getopt.h>` on line 156, which I did in https://github.com/eecs280staff/tutorials/pull/40/commits/39738b5dfb80e927a376971aa6e1e69f5ba10d34

Reply:
Good catch!  OK, I simplified EVEN MORE!  I moved the `#ifdef _WIN32` stuff inside the .h file.  I also renamed it to `xgetopt.h` because it cross platform (Sometimes we use "X" to mean "cross platform".). I also updated the instructions to be all in one place now that they are vastly simpler.  Try this version!

Reply:
The new `xgetopt.h` version is working for me locally (linux) and on the 281 AG, but I'm having the same issues that I mentioned here:

> VS is giving compilation issues on this line saying that a const char * cant be cast to a char *: ![image](https://user-images.githubusercontent.com/70495470/205176621-1a9c66e8-8708-46d0-8519-bf3bf4ae4dd6.png)
> 

On VS.

Reply:
Hmm, I thought the fix was the comment out this line, which I did: `//#include <getopt.h>`.  Did I miss part of your recommended fix?

Reply:
Sorry for the delay in responding. 

When I compiled with the new `xgetopt.h` file, I got both a compilation error for the `#include <getopt.h>`, as well as, the `const char *` -> `char *` conversion error.

Commenting out the  `#include <getopt.h>`, got rid of the first error, but I'm still experiencing the  `const char *` -> `char *` conversion error.

VS is only saying that [`strchr`](https://cplusplus.com/reference/cstring/strchr/) will only return a `const char *`, but the C++ reference says it is overloaded to return either a `char *` or a `const char *`. We need `strchr` to return a `char *` for it to work.

A (less than ideal) fix would be to implement our own version of `strchr`, which wouldn't be too hard considering that all `strchr` does is "Returns a pointer to the first occurrence of character in the C string str."

I've implemented a version of `strchr` called `mutable_strchr` in https://github.com/eecs280staff/tutorials/pull/40/commits/0113e034c03ccd953d8cc4d6a69a2d911a0acd37, and I've verified locally that it works.

This eliminates both of the errors that I was seeing locally.